File: src/python/networks/rnn/rnn.py
from numba import jit
import numpy as np

from utils.optimizers.adagrad import Adagrad
from utils.optimizers.gradient_descent import GradientDescent
from utils.optimizers.adam import Adam
from utils.layers.dense import Layer, GPULayer
from utils.layers.lstm import LSTMCell
try:
    from utils.cuda.cuda import gpu_mse_loss
except ImportError:
    def gpu_mse_loss(pred, target, N):
        return ((pred - target) ** 2).mean()


class RNN:
    def __init__(self, layers_config, optimizer_name='adam', optimizer_params=None, **kwargs):
        use_cuda = kwargs.get('cuda', False)
        self._loss_function = kwargs.get('loss_function', 'mse')

        loss_functions = self._init_loss_functions()
        self._loss = loss_functions[0]
        self._loss_grad = loss_functions[1]

        self.layers = []

        for config in layers_config:
            try:
                layer_type = config['type']
            except KeyError:
                layer_type = 'dense'

            if layer_type == 'lstm':
                if use_cuda:
                    raise NotImplementedError("LSTM not implemented for GPU")
                print(f"Initializing LSTMCell with input_size={config['input_size']} and hidden_size={config['output_size']}")
                self.layers.append(LSTMCell(config['input_size'], config['output_size']))
            else:
                if use_cuda:
                    print(f"Initializing GPU Layers with input_size={config['input_size']} and output_size={config['output_size']}")
                    self.layers.append(GPULayer(config['input_size'], config['output_size'], config['activation']))
                else:
                    print(f"Initializing Layer with input_size={config['input_size']} and output_size={config['output_size']}")
                    self.layers.append(Layer(config['input_size'], config['output_size'], config['activation']))


        if optimizer_params is None:
            optimizer_params = {'lr': 0.01}

        self.optimizer = self._create_optimizer(optimizer_name, optimizer_params)

    def _create_optimizer(self, optimizer_name, optimizer_params):
        name = optimizer_name.lower()
        if name == 'gradientdescent':
            optimizer = GradientDescent(**optimizer_params)
        elif name == 'adagrad':
            optimizer = Adagrad(**optimizer_params)
            param_shapes = {f'layer_{i}_weights': layer.weights.shape for i, layer in enumerate(self.layers)}
            param_shapes.update({f'layer_{i}_biases': layer.biases.shape for i, layer in enumerate(self.layers)})
            optimizer.initialize_parameters(param_shapes)
        elif name == 'adam':
            optimizer = Adam(**optimizer_params)
            param_shapes = {f'layer_{i}_weights': layer.weights.shape for i, layer in enumerate(self.layers)}
            param_shapes.update({f'layer_{i}_biases': layer.biases.shape for i, layer in enumerate(self.layers)})
            optimizer.initialize_parameters(param_shapes)
        else:
            raise ValueError(f"Unsupported optimizer: {optimizer_name}")
        return optimizer

    def forward(self, x):
        batch_size = x.shape[1]
        hidden_states = [(np.zeros((layer.hidden_size, batch_size)), np.zeros((layer.hidden_size, batch_size))) 
                        for layer in self.layers if isinstance(layer, LSTMCell)]

        h_c_index = 0
        for i, layer in enumerate(self.layers):
            if isinstance(layer, LSTMCell):
                h, c = hidden_states[h_c_index]
                h, c = layer.forward(x, h, c)
                hidden_states[h_c_index] = (h, c)
                x = h  # Update x to the output h for the next layer
                h_c_index += 1
            else:
                x = layer.forward(x)
        return x

    def backward(self, dh):
        dc = np.zeros_like(dh)  # Only needed if the last layer is LSTM
        for i, layer in reversed(list(enumerate(self.layers))):
            if isinstance(layer, LSTMCell):
                dx, dh, dc = layer.backward(dh, dc)
            else:
                # Ensure dh has the correct shape for dense layers
                if dh.shape[0] != layer.output_size:
                    dh = np.resize(dh, (layer.output_size, dh.shape[1]))
                dh = layer.backward(dh)

    def adjust_gradients(self, dh, dc, hidden_size):
        # Resize gradients to match the hidden size of the next layer in the backward pass
        if dh.shape[0] != hidden_size:
            dh = np.resize(dh, (hidden_size, dh.shape[1]))
            dc = np.resize(dc, (hidden_size, dc.shape[1]))
        return dh, dc

    def update_weights(self):
        for i, layer in enumerate(self.layers):
            layer.weights = self.optimizer.update(layer.weights, layer.grad_weights, f'layer_{i}_weights')
            layer.biases = self.optimizer.update(layer.biases, layer.grad_biases, f'layer_{i}_biases')

    def train(self, inputs, targets, cuda=False):
        if not cuda:
            outputs = self.forward(inputs)
            loss = self._loss(outputs, targets)
            grad_outputs = self._loss_grad(outputs, targets)
            self.backward(grad_outputs)
            self.update_weights()
            return loss
        else:
            return self._train_gpu(inputs, targets)

    def _train_gpu(self, inputs, targets):
        outputs = self.forward(inputs)
        loss = gpu_mse_loss(outputs, targets, outputs.size)
        grad_outputs = 2 * (outputs - targets) / outputs.size
        self.backward(grad_outputs)
        self.update_weights()
        return loss
    
    def _init_loss_functions(self):
        loss_function = self._loss_function.lower()
        if loss_function == 'mse':
            return self.mean_squared_error, self.mse_grad
        elif loss_function == 'crossentropy':
            return self.cross_entropy_loss, self.cross_entropy_grad
        else:
            raise ValueError(f"Unsupported loss function: {self._loss_function}")

    @staticmethod
    @jit(nopython=True, fastmath=True, cache=True)
    def mean_squared_error(outputs, targets):
        return ((outputs - targets) ** 2).mean()
    
    @staticmethod
    @jit(nopython=True, fastmath=True, cache=True)
    def mse_grad(outputs, targets):
        return 2 * (outputs - targets) / outputs.size
    
    @staticmethod
    @jit(nopython=True, fastmath=True, cache=True)
    def cross_entropy_loss(outputs, targets):
        return -np.sum(targets * np.log(outputs)) / outputs.shape[1]
    
    @staticmethod
    @jit(nopython=True, fastmath=True, cache=True)
    def cross_entropy_grad(outputs, targets):
        return outputs - targets

    @jit(nopython=False, fastmath=True, cache=True, nogil=True)
    def predict(self, x):
        predictions = []
        for i in range(len(x)):
            prediction = self.forward(x[i].reshape(-1, 1))
            predictions.append(prediction.flatten())

        return np.array(predictions)

File: src/rust/src/rnn.rs
use ndarray::{Array1, Array2, Axis, stack};
use std::collections::HashMap;

use crate::layers::{Layer, dense::DenseLayer, lstm::LSTMCell};
use crate::optimizer::{Optimizer, adam::Adam, gradient_descent::GradientDescent};

pub struct RNN {
    layers: Vec<Box<dyn Layer>>,
    optimizer: Box<dyn Optimizer>,
    loss_function: LossFunction,
}

impl RNN {
    pub fn new(layers_config: Vec<LayerConfig>, optimizer_name: &str, optimizer_params: Option<HashMap<String, f64>>, cuda: bool, loss_function_name: &str) -> Self {
        let loss_function = match loss_function_name {
            "mse" => LossFunction::MSE,
            "crossentropy" => LossFunction::CrossEntropy,
            _ => panic!("Unsupported loss function: {}", loss_function_name),
        };

        let mut layers: Vec<Box<dyn Layer>> = Vec::new();

        for config in layers_config {
            match config.layer_type.as_str() {
                "lstm" => {
                    if cuda {
                        panic!("LSTM not implemented for GPU");
                    }
                    println!("Initializing LSTMCell with input_size={} and hidden_size={}", config.input_size, config.output_size);
                    layers.push(Box::new(LSTMCell::new(config.input_size, config.output_size)));
                }
                _ => {
                    if cuda {
                        println!("Initializing GPU Layers with input_size={} and output_size={}", config.input_size, config.output_size);
                        // Replace GPULayer with the correct struct if it's implemented
                        // layers.push(Box::new(GPULayer::new(config.input_size, config.output_size, config.activation.clone())));
                    } else {
                        println!("Initializing Layer with input_size={} and output_size={}", config.input_size, config.output_size);
                        layers.push(Box::new(DenseLayer::new(config.input_size, config.output_size, config.activation.clone())));
                    }
                }
            }
        }

        let optimizer_params = optimizer_params.unwrap_or_else(|| {
            let mut default_params = HashMap::new();
            default_params.insert("lr".to_string(), 0.01);
            default_params
        });

        let mut optimizer = RNN::create_optimizer(optimizer_name, &optimizer_params);

        // Collect parameter shapes
        let param_shapes = collect_param_shapes(&layers);
        optimizer.initialize_parameters(&param_shapes);

        RNN {
            layers,
            optimizer,
            loss_function,
        }
    }

    fn create_optimizer(name: &str, params: &HashMap<String, f64>) -> Box<dyn Optimizer> {
        match name.to_lowercase().as_str() {
            "gradientdescent" => Box::new(GradientDescent::new(params)),
            "adam" => {
                let optimizer = Adam::new(params);
                Box::new(optimizer)
            }
            _ => panic!("Unsupported optimizer: {}", name),
        }
    }

    pub fn forward(&mut self, x: Array2<f64>) -> Array2<f64> {
        let mut x = x;
        let mut hidden_states: Vec<(Array2<f64>, Array2<f64>)> = Vec::new();
    
        // Initialize hidden states
        for layer in &self.layers {
            if let Some(lstm_layer) = layer.as_any().downcast_ref::<LSTMCell>() {
                let h = Array2::zeros((lstm_layer.hidden_size(), x.ncols()));
                let c = Array2::zeros((lstm_layer.hidden_size(), x.ncols()));
                hidden_states.push((h, c));
            }
        }
    
        let mut h_c_index = 0;
        for layer in &mut self.layers {  // Mutable reference here
            if let Some(lstm_layer) = layer.as_any_mut().downcast_mut::<LSTMCell>() { // downcast_mut
                let (mut h, mut c) = hidden_states[h_c_index].clone();
                let (new_h, new_c) = lstm_layer.forward(&x, &mut h, &mut c);
                hidden_states[h_c_index] = (new_h.clone(), new_c.clone());
                x = new_h;
                h_c_index += 1;
            } else {
                x = layer.forward(&x);
            }
        }
    
        x
    }    

    pub fn backward(&mut self, dh: Array2<f64>) {
        let mut dh = dh;
        let mut dc = Array2::zeros(dh.dim()); // Only needed if the last layer is LSTM
    
        for (_i, layer) in self.layers.iter_mut().enumerate().rev() { // Mutable iteration
            if let Some(lstm_layer) = layer.as_any_mut().downcast_mut::<LSTMCell>() { // Mutable downcast
                let (new_dh, new_dc, _new_dc_prev) = lstm_layer.backward(&dh, &dc);
                dh = new_dh;
                dc = new_dc;
            } else {
                if dh.nrows() != layer.output_size() {
                    dh = dh.clone().into_shape((layer.output_size(), dh.ncols())).unwrap();
                }
                dh = layer.backward(&dh);
            }
        }
    }
    
    pub fn update_weights(&mut self) {
        for (i, layer) in self.layers.iter_mut().enumerate() {
            layer.update_weights(&mut *self.optimizer, i);
        }
    }

    pub fn train(&mut self, inputs: Array2<f64>, targets: Array2<f64>, cuda: bool) -> f64 {
        if !cuda {
            let outputs = self.forward(inputs);
            let loss = self.loss(outputs.clone(), targets.clone());
            let grad_outputs = self.loss_grad(outputs, targets);
            self.backward(grad_outputs);
            self.update_weights();
            loss
        } else {
            self.train_gpu(inputs, targets)
        }
    }

    fn train_gpu(&mut self, inputs: Array2<f64>, targets: Array2<f64>) -> f64 {
        let outputs = self.forward(inputs);
        let loss = gpu_mse_loss(outputs.clone(), targets.clone());
        let grad_outputs = 2.0 * (outputs.clone() - targets) / outputs.len_of(Axis(1)) as f64;
        self.backward(grad_outputs);
        self.update_weights();
        loss
    }

    fn loss(&self, outputs: Array2<f64>, targets: Array2<f64>) -> f64 {
        match self.loss_function {
            LossFunction::MSE => mean_squared_error(outputs, targets),
            LossFunction::CrossEntropy => cross_entropy_loss(outputs, targets),
        }
    }

    fn loss_grad(&self, outputs: Array2<f64>, targets: Array2<f64>) -> Array2<f64> {
        match self.loss_function {
            LossFunction::MSE => mse_grad(outputs, targets),
            LossFunction::CrossEntropy => cross_entropy_grad(outputs, targets),
        }
    }

    pub fn predict(&mut self, x: Array2<f64>) -> Array2<f64> {
        let mut predictions: Vec<Array1<f64>> = Vec::new();
    
        for row in x.rows() {
            let input = row.to_owned().insert_axis(Axis(1));
            let prediction = self.forward(input);
            predictions.push(prediction.clone().into_shape(prediction.len()).unwrap());
        }
    
        stack(Axis(0), &predictions.iter().map(|a| a.view()).collect::<Vec<_>>()).unwrap()
    }    
}

#[derive(Clone)]
pub struct LayerConfig {
    pub layer_type: String,
    pub input_size: usize,
    pub output_size: usize,
    pub activation: Option<String>,
}

pub enum LossFunction {
    MSE,
    CrossEntropy,
}

pub fn mean_squared_error(outputs: Array2<f64>, targets: Array2<f64>) -> f64 {
    ((&outputs - &targets).mapv(|x| x.powi(2)).sum()) / outputs.len_of(Axis(1)) as f64
}

pub fn mse_grad(outputs: Array2<f64>, targets: Array2<f64>) -> Array2<f64> {
    2.0 * (outputs.clone() - targets) / outputs.len_of(Axis(1)) as f64
}

pub fn cross_entropy_loss(outputs: Array2<f64>, targets: Array2<f64>) -> f64 {
    -(&targets * outputs.mapv(|x| x.ln())).sum() / outputs.len_of(Axis(1)) as f64
}

pub fn cross_entropy_grad(outputs: Array2<f64>, targets: Array2<f64>) -> Array2<f64> {
    outputs - targets
}

pub fn gpu_mse_loss(outputs: Array2<f64>, targets: Array2<f64>) -> f64 {
    mean_squared_error(outputs, targets) // Placeholder for actual GPU implementation
}

fn collect_param_shapes(layers: &Vec<Box<dyn Layer>>) -> HashMap<String, (usize, usize)> {
    let mut param_shapes = HashMap::new();

    for (i, layer) in layers.iter().enumerate() {
        let layer_name = format!("layer_{}", i);
        if let Some(dense_layer) = layer.as_any().downcast_ref::<DenseLayer>() {
            param_shapes.insert(format!("{}_weights", layer_name), (dense_layer.output_size(), dense_layer.input_size()));
            param_shapes.insert(format!("{}_biases", layer_name), (dense_layer.output_size(), 1));
        } else if let Some(lstm_layer) = layer.as_any().downcast_ref::<LSTMCell>() {
            param_shapes.insert(format!("{}_weights", layer_name), (4 * lstm_layer.hidden_size(), lstm_layer.input_size() + lstm_layer.hidden_size()));
            param_shapes.insert(format!("{}_biases", layer_name), (4 * lstm_layer.hidden_size(), 1));
        }
    }

    param_shapes
}

